### ISSUE TITLE
_cleanup(None): invalid exception object

### ISSUE THREAD

**Issue by HynekPetrak** [ronf/asyncssh#185]
Sometimes I get the below error. How can I debug that?
```
Exception in callback SSHClientConnection._cleanup(None)
handle: <Handle SSHClientConnection._cleanup(None)>
Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "/root/sshame/venv/lib/python3.6/site-packages/asyncssh-1.15.0-py3.6.egg/asyncssh/connection.py", line 2289, in _cleanup
    self._auth_waiter.set_exception(exc)
TypeError: invalid exception object
```

**Comment by ronf** [comment]
The error appears like it could be caused by getting into SSHClientConnection:_cleanup() with an exception value of `None` prior to authentication completing. Most code paths turn `None` into a DisconnectError(), which would avoid this error being generated. However, it looks like there may be a code path where _cleanup() is called as a result of the remote server sending back a MSG_DISCONNECT with a code of DISC_BY_APPLICATION that might trigger this. This message would normally be used for a clean close, though, and I wouldn't expect that to be sent prior to authentication completing.

What are you using as a server here? If you are using AsyncSSH on both sides, are you calling close() or disconnect() on the server side in some cases before authentication is complete on the client?

It looks like it might be possible to fix this by moving the code in connection_lost() of:

```python
        if exc is None and self._transport:
            exc = DisconnectError(DISC_CONNECTION_LOST, 'Connection lost')
```

into _force_close() instead, so the eventual call to _cleanup() never receives `None` as a value for exc. I'd need to do a bit of testing to make sure that this didn't cause problems in debug logging for clean closes, though - I've run into problems in the past with that, which is why `None` was being passed in some cases.

The other option would be to put something in SSHClientConnection:_cleanup() itself to test if exc is None and fill in something like the above DisconnectError() in the call to set_exception(), but I'd rather not duplicate that logic there.

Do you have a simple test case which produces this error? Alternately, seeing output from the debug logging might help to confirm what's happening. I think the default of debug level 1 is probably good enough. You'd just need to add something like the following in your code before you make the SSH client connection:

```python
import logging
logging.basicConfig(level='DEBUG')
```

**Comment by ronf** [comment]
I spent a bit more time on this and was able to reproduce this error by having the server send a disconnect message of code DISC_BY_APPLICATION before authentication was complete. Here's a patch I'm looking at applying to the "develop" branch which addresses both this issue and an exception on the server side which occurred when attempting to close the connection from within an auth callback:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index f15d1aa..f820bdf 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -1234,7 +1234,7 @@ class SSHConnection(SSHPacketHandler):
 
         self.logger.debug1('Received disconnect: %s (%d)', reason, code)
 
-        if code != DISC_BY_APPLICATION:
+        if code != DISC_BY_APPLICATION or self._auth_waiter:
             exc = DisconnectError(code, reason, lang)
         else:
             exc = None
@@ -1507,6 +1507,9 @@ class SSHConnection(SSHPacketHandler):
     def _finish_userauth(self, result, method, packet):
         """Finish processing a user authentication request"""
 
+        if not self._owner:
+            return
+
         if asyncio.iscoroutine(result):
             result = yield from result

```

The change makes it so that any disconnects during auth will report back the full exception, including the string sent by the server in its disconnect message. However, after authentication is complete, any DISC_BY_APPLICATION disconnect messages sent in either direction will be reported as a clean connection close (as they did before), since those are generally not an indication of an error.

If all goes well after some additional testing, I'll check this into the "develop" branch and plan on rolling this fix into the next release. Thanks for the report!

**Comment by HynekPetrak** [comment]
Hello Ron,
the patch seems to be correct for my case. I'm not getting the error anymore. 
Thanks a lot!

Btw. what's the purpose of the "else" branch? It might case the same issue or not?
```
+        if code != DISC_BY_APPLICATION or self._auth_waiter:
             exc = DisconnectError(code, reason, lang)
         else:
             exc = None
```

**Comment by ronf** [comment]
This is now checked into the develop branch as commit ecf5ed3 and will be included in the next release.

Regarding the else clause, `None` is used to indicate a clean close, which reports connection close to the application code with raising any exceptions. The crash you ran into happened because we were trying to call set_exception() on the `Future` that the AsyncSSH connect() call was blocked on, and that required a real exception to be delivered. The new condition which looks to see if self._auth_waiter is set now makes sure that a real exception is returned in that case. Once auth is complete, though, the previous code path where `None` was used works fine, as it doesn't actually call set_exception() on anything.

**Comment by ronf** [comment]
This fix is now available in 1.16.0.
